core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -158,22 +161,31 @@ public List<Delete> getDeletesInDeleteSet() {
     return new ArrayList<>(deleteSet.values());
   }
 
-  public Optional<TransactionResult> get(Key key) throws CrudException {
+  public Optional<TransactionResult> mergeResult(Key key, Optional<TransactionResult> result)
+      throws CrudException {
     if (deleteSet.containsKey(key)) {
       return Optional.empty();
-    } else if (readSet.containsKey(key)) {
-      if (writeSet.containsKey(key)) {
-        // merge the result in the read set and the put in the write set
-        return Optional.of(
-            new TransactionResult(
-                new MergedResult(readSet.get(key), writeSet.get(key), getTableMetadata(key))));
-      } else {
-        return readSet.get(key);
-      }
+    } else if (writeSet.containsKey(key)) {
```

Comment:
Can I ask an additional question?
https://github.com/scalar-labs/scalardb/pull/1976/files#diff-4e6eb5a222519a33681ef588e24c3a8eabc8bdec7e89e79bb5c96c3ce8898759R143-R146 isn't atomic operation, so this method could potentially return `result` from the last `else` block. Is it expected?

Reply:
@komamitsu Good point, thanks. I don't think the basic behavior is changed from the previous one.
https://github.com/scalar-labs/scalardb/blob/97aabbfd6d0da98e84ff45af0d0f78d88f60e1b7/core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java#L162-L172

So, it could return an unexpected (neither merged nor deleted) result from the cache even though it is overwritten and deleted. I don't think it's critical since it's rare to execute operations in parallel in the same transaction if we say "Snapshot is not thread-safe" in the document.

However, if we change the order like this, I guess it will be improved since we check the delete set first. What do you think?

```java
  public void put(Key key, Delete delete) {
    deleteSet.put(key, delete);
    writeSet.remove(key);
  }
```

Reply:
> I don't think it's critical since it's rare to execute operations in parallel in the same transaction if we say "Snapshot is not thread-safe" in the document.

Yes, we don't assume users will execute operations concurrently within a single transaction. So, the `Snapshot` class is basically not thread-safe.

Reply:
> However, if we change the order like this, I guess it will be improved since we check the delete set first. What do you think?

Yeah, I think it's better. But I guess there is still a chance that something wrong could occur?

- Thread A enters `mergeResult()` and checks `deleteSet`. It doesn't find the key
- Thread B enters `put()`, and executes `deleteSet.put(key, delete)` and `writeSet.remove(key)`
- Thread A checks `writeSet`, but doesn't find the key. It moves to the last `else` block

Also, `deleteSet` and `writeSet` are HashMap, so I guess some invisible update issue might happen due to lack up memory barrier.

But, as you guys said, I understand concurrent multiple CRUD operations don't make sense. So, I think we don't need to consider that. Thanks for the replies!

Reply:
@komamitsu @brfrn169 OK, I will leave it as is. Thank you for the discussion.